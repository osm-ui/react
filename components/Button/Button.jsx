import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styled from 'styled-components';


const contexts = [
    'default',
    'primary',
    'info',
    'success',
    'warning',
    'danger',
    'link',
];

const colorsStyle = props => contexts.reduce((reducedStyles, context) => {
    const colors = context === 'default'
        ? props.theme.form.button
        : props.theme.form.button[context];

    return (`
        ${reducedStyles}

        &.btn-${context} {
            font-weight: ${colors.fontWeight};
            color: ${colors.color};
            background-color: ${colors.backgroundColor};
            border-color: ${colors.borderColor};

            &:hover {
                color: ${colors.color};
                background-color: ${colors.hoverBackgroundColor};
                border-color: ${colors.hoverBorderColor};
            }

            &:focus {
                color: ${colors.color};
                background-color: ${colors.focusBackgroundColor};
                border-color: ${colors.focusBorderColor};
            }

            &:active, .active {
                color: ${colors.color};
                background-color: ${colors.activeBackgroundColor};
                border-color: ${colors.activeBorderColor};
            }
        }
    `);
}, '');


const StyledButton = styled.button`
    ${props => colorsStyle(props)}
`;

const StyledAnchor = styled.a`
    ${colorsStyle}
`;


const Button = ({
    type,
    context,
    size,
    block,
    active,
    disabled,
    ...props
}) => {
    const classes = classnames({
        btn: true,
        [`btn-${context}`]: true,
        [`btn-${size}`]: true,
        'btn-block': block,
        active,
        disabled,
    });

    if (type === 'anchor') {
        return (
            <StyledAnchor className={classes} role="button" {...props} />
        );
    }

    return (
        <StyledButton className={classes} type={type} disabled={disabled} {...props} />
    );
};


Button.propTypes = {
    type: PropTypes.string,
    context: PropTypes.oneOf(contexts),
    size: PropTypes.oneOf([
        'lg',
        'md',
        'sm',
        'xs',
    ]),
    block: PropTypes.bool,
    active: PropTypes.bool,
    disabled: PropTypes.bool,
};

Button.defaultProps = {
    type: 'button',
    context: 'default',
    size: 'md',
    block: false,
    active: false,
    disabled: false,
};

Button.displayName = 'Button';

export default Button;
