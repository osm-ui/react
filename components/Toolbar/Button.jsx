import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';
import FontAwesome from 'react-fontawesome';
import Loader from '../Loader';


const StyledButton = styled.button`
    position: relative;
    transition: all 0.1s ease-out;

    color: ${props => props.theme.toolbar.button.color};
    border-color: ${props => props.theme.toolbar.button.borderColor};
    border-style: ${props => props.theme.toolbar.button.borderStyle};
    border-width: ${props => props.theme.toolbar.button.borderWidth};
    background: ${props => props.theme.toolbar.button.backgroundColor};
    padding: 0;

    &:hover {
        color: ${props => props.theme.toolbar.button.color};
        background-color: ${props => props.theme.toolbar.button.hoverBackgroundColor};
        border-color: ${props => props.theme.toolbar.button.hoverBorderColor};
    }

    &:focus {
        color: ${props => props.theme.toolbar.button.color};
        background-color: ${props => props.theme.toolbar.button.focusBackgroundColor};
        border-color: ${props => props.theme.toolbar.button.focusBorderColor};
    }

    &:active, .active {
        color: ${props => props.theme.toolbar.button.color};
        background-color: ${props => props.theme.toolbar.button.activeBackgroundColor};
        border-color: ${props => props.theme.toolbar.button.activeBorderColor};
    }

    &.shape-square {
        border-radius: 0;
    }

    &.shape-round {
        border-radius: 50%;
    }

    &.xs {
        width: ${props => props.theme.toolbar.xsSize};
        height: ${props => props.theme.toolbar.xsSize};
        font-size: 0.85em;
        line-height: 0.85em;
    }

    &.sm {
        width: ${props => props.theme.toolbar.smSize};
        height: ${props => props.theme.toolbar.smSize};
        font-size: 1.08em;
        line-height: 1.08em;
    }

    &.md {
        width: ${props => props.theme.toolbar.mdSize};
        height: ${props => props.theme.toolbar.mdSize};
        font-size: 1.29em;
        line-height: 1.29em;
    }

    &.lg {
        width: ${props => props.theme.toolbar.lgSize};
        height: ${props => props.theme.toolbar.lgSize};
        font-size: 1.43em;
        line-height: 1.43em;
    }
`;


const ToolbarButton = ({
    // title,
    icon,
    size,
    shape,
    loading,
    // loaderTitle,
    className,
    children,
    ...rest
}) => {
    const classes = classnames(className, {
        loading,
        [size]: size,
        [`shape-${shape}`]: true,
        btn: true,
    });

    if (loading) {
        const spinnerSizes = {
            xs: 14,
            sm: 20,
            md: 24,
            lg: 30,
        };
        const strokeSizes = {
            xs: 2,
            sm: 3,
            md: 3,
            lg: 3,
        };
        return (
            <StyledButton
                className={classes}
                {...rest}
            >
                <Loader
                    spinnerSize={spinnerSizes[size]}
                    strokeSize={strokeSizes[size]}
                    centered
                />
            </StyledButton>
        );
    }

    return (
        <StyledButton
            className={classes}
            {...rest}
        >
            {children && children}
            {!children && <FontAwesome name={icon} />}
        </StyledButton>
    );
};


ToolbarButton.propTypes = {
    // title: PropTypes.string,
    icon: PropTypes.string,
    size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
    shape: PropTypes.oneOf(['round', 'square']),
    loading: PropTypes.bool,
    // loaderTitle: PropTypes.node,
    className: PropTypes.string,
    children: PropTypes.node,
};

ToolbarButton.defaultProps = {
    // title: '',
    icon: '',
    size: 'md',
    shape: 'round',
    loading: false,
    // loaderTitle: '',
    className: '',
    children: '',
};

ToolbarButton.displayName = 'Toolbar.Button';

export default ToolbarButton;
