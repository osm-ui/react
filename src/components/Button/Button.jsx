import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styled from 'styled-components';

const contexts = [
  'default',
  'primary',
  'info',
  'success',
  'warning',
  'danger',
  'link'
];

const colorsStyle = props =>
  contexts.reduce((reducedStyles, context) => {
    const colors =
      context === 'default'
        ? props.theme.form.button
        : props.theme.form.button[context];

    return `
        ${reducedStyles}

        &.btn-${context} {
            font-weight: ${colors.fontWeight};
            color: ${colors.color};
            background-color: ${colors.backgroundColor};
            border-color: ${colors.borderColor};

            &:hover {
                color: ${colors.color};
                background-color: ${colors.hoverBackgroundColor};
                border-color: ${colors.hoverBorderColor};
            }

            &:focus {
                color: ${colors.color};
                background-color: ${colors.focusBackgroundColor};
                border-color: ${colors.focusBorderColor};
            }

            &:active, .active {
                color: ${colors.color};
                background-color: ${colors.activeBackgroundColor};
                border-color: ${colors.activeBorderColor};
            }
        }
    `;
  }, '');

const StyledButton = styled.button`
  ${props => colorsStyle(props)} border-radius: ${props =>
      props.theme.form.button.borderRadius};

  &.shape-round {
    &.btn-xs {
      height: 25px;
      border-radius: 12.5px;
      padding: 1px 14px;
    }

    &.btn-sm {
      height: 33px;
      border-radius: 16.5px;
      padding: 5px 16px;
    }

    &.btn-md {
      height: 42px;
      border-radius: 21px;
      padding: 6px 20px;
    }

    &.btn-lg {
      height: 49px;
      border-radius: 24.5px;
      padding: 10px 24px;
    }
  }
`;

const StyledAnchor = StyledButton.withComponent('a');

const Button = ({
  type,
  context,
  shape,
  size,
  block,
  active,
  disabled,
  className,
  ...props
}) => {
  const classes = classnames(className, {
    btn: true,
    [`shape-${shape}`]: true,
    [`btn-${context}`]: true,
    [`btn-${size}`]: true,
    'btn-block': block,
    active,
    disabled
  });

  if (type === 'anchor') {
    return (
      <StyledAnchor
        className={classes}
        disabled={disabled}
        role="button"
        {...props}
      />
    );
  }

  return (
    <StyledButton
      className={classes}
      type={type}
      disabled={disabled}
      {...props}
    />
  );
};

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit', 'reset', 'anchor']),
  context: PropTypes.oneOf(contexts),
  shape: PropTypes.oneOf(['square', 'round']),
  size: PropTypes.oneOf(['lg', 'md', 'sm', 'xs']),
  block: PropTypes.bool,
  active: PropTypes.bool,
  disabled: PropTypes.bool,
  className: PropTypes.string
};

Button.defaultProps = {
  type: 'button',
  context: 'default',
  shape: 'round',
  size: 'md',
  block: false,
  active: false,
  disabled: false,
  className: ''
};

Button.displayName = 'Button';

export default Button;
