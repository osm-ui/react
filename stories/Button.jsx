import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { host } from 'storybook-host';
import { withKnobs, text, boolean, select } from '@storybook/addon-knobs';

import defaultHostOptions from './defaultHostOptions';
import KnobsAlert from './components/KnobsAlert';

import {
  DefaultTheme,
  WhiteTheme,
  LightGrayTheme,
  DarkGrayTheme,
  AnthraciteTheme,
  YellowTheme,
  OrangeTheme,
  BrownTheme,
  RedTheme,
  RoseTheme,
  PurpleTheme,
  BlueTheme,
  SkyTheme,
  TurquoiseTheme,
  GreenTheme,
  Section,
  Button
} from 'index';

storiesOf('Button', module)
  .addDecorator(withKnobs)
  .addDecorator(
    host({
      ...defaultHostOptions,
      title: 'Button'
    })
  )
  .addWithInfo('Context', () => (
    <DefaultTheme>
      <Button context="default" onClick={action('onClick default')}>
        Default
      </Button>{' '}
      <Button context="primary" onClick={action('onClick primary')}>
        Primary
      </Button>{' '}
      <Button context="success" onClick={action('onClick success')}>
        Success
      </Button>{' '}
      <Button context="info" onClick={action('onClick info')}>
        Info
      </Button>{' '}
      <Button context="warning" onClick={action('onClick warning')}>
        Warning
      </Button>{' '}
      <Button context="danger" onClick={action('onClick danger')}>
        Danger
      </Button>{' '}
      <Button context="link" onClick={action('onClick link')}>
        Link
      </Button>
    </DefaultTheme>
  ))
  .addWithInfo('Type', () => (
    <DefaultTheme>
      <Button type="button" onClick={action('onClick button')}>
        Button (default)
      </Button>{' '}
      <Button type="submit" onClick={action('onClick submit')}>
        Submit button
      </Button>{' '}
      <Button type="reset" onClick={action('onClick reset')}>
        Reset button
      </Button>{' '}
      <Button type="anchor" href="#" onClick={action('onClick anchor')}>
        Anchor button
      </Button>
    </DefaultTheme>
  ))
  .addWithInfo('Shape', () => (
    <DefaultTheme>
      <Button shape="round" onClick={action('onClick round')}>
        Round (default)
      </Button>{' '}
      <Button shape="square" onClick={action('onClick square')}>
        Square
      </Button>
    </DefaultTheme>
  ))
  .addWithInfo('Size', () => (
    <DefaultTheme>
      <Button size="lg" onClick={action('onClick lg')}>
        Large button
      </Button>{' '}
      <Button size="md" onClick={action('onClick md')}>
        Medium button (default)
      </Button>{' '}
      <Button size="sm" onClick={action('onClick sm')}>
        Small button
      </Button>{' '}
      <Button size="xs" onClick={action('onClick xs')}>
        Extra small button
      </Button>
    </DefaultTheme>
  ))
  .addWithInfo('Block', () => (
    <DefaultTheme>
      <Button
        block
        size="lg"
        context="primary"
        onClick={action('onClick primary')}
      >
        Block level button
      </Button>
      <Button block size="lg" onClick={action('onClick default')}>
        Block level button
      </Button>
    </DefaultTheme>
  ))
  .addWithInfo('State', () => (
    <DefaultTheme>
      <Button active onClick={action('onClick')}>
        Active
      </Button>{' '}
      <Button disabled>Disabled</Button>
    </DefaultTheme>
  ))
  .addWithInfo('Playground', () => {
    const label = text('Label', 'Hello world!');
    const block = boolean('Block');
    const active = boolean('Active');
    const disabled = boolean('Disabled');
    const shape = select('Shape', ['round', 'square'], 'round');
    const size = select('Size', ['lg', 'md', 'sm', 'xs'], 'md');
    const type = select(
      'Type',
      ['button', 'submit', 'reset', 'anchor'],
      'button'
    );
    const context = select(
      'Context',
      ['default', 'info', 'primary', 'success', 'warning', 'danger'],
      'default'
    );
    const theme = select(
      'Theme',
      [
        'Default',
        'White',
        'LightGray',
        'DarkGray',
        'Anthracite',
        'Yellow',
        'Orange',
        'Brown',
        'Red',
        'Rose',
        'Purple',
        'Blue',
        'Sky',
        'Turquoise',
        'Green'
      ],
      'Default'
    );
    const themes = {
      DefaultTheme,
      WhiteTheme,
      LightGrayTheme,
      DarkGrayTheme,
      AnthraciteTheme,
      YellowTheme,
      OrangeTheme,
      BrownTheme,
      RedTheme,
      RoseTheme,
      PurpleTheme,
      BlueTheme,
      SkyTheme,
      TurquoiseTheme,
      GreenTheme
    };
    const ThemeElement = themes[`${theme}Theme`];

    return (
      <DefaultTheme>
        <KnobsAlert />
        <ThemeElement>
          <Section style={{ padding: 30, display: 'inline' }}>
            <Button
              block={block}
              active={active}
              disabled={disabled}
              shape={shape}
              size={size}
              context={context}
              type={type}
              onClick={action('onClick')}
            >
              {label}
            </Button>
          </Section>
        </ThemeElement>
      </DefaultTheme>
    );
  });
