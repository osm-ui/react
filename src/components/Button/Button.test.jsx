import React from 'react';
import {
  snapshotWithElementChildren,
  snapshotWithTextChildren
} from '../../helpers/tests';
import Button from './';

describe('When using snapshots', () => {
  it('Should render with an element children', () =>
    snapshotWithElementChildren(Button));

  it('Should render with a text children', () =>
    snapshotWithTextChildren(Button));
});

describe('When using snapshots on an anchor typed Button', () => {
  it('Should render with an element children', () =>
    snapshotWithElementChildren(Button, { type: 'anchor' }));

  it('Should render with a text children', () =>
    snapshotWithTextChildren(Button, { type: 'anchor' }));
});
