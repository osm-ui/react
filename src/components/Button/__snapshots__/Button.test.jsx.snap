// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When using snapshots Should render with a text children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <button
    className="btn shape-round btn-default btn-md sc-bwzfXH fEgsfH"
    disabled={false}
    type="button"
  >
    This is a text
  </button>
</span>
`;

exports[`When using snapshots Should render with an element children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <button
    className="btn shape-round btn-default btn-md sc-bwzfXH fEgsfH"
    disabled={false}
    type="button"
  >
    <div />
  </button>
</span>
`;

exports[`When using snapshots on an anchor typed Button Should render with a text children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <a
    className="btn shape-round btn-default btn-md sc-htpNat jTMgHP"
    disabled={false}
    role="button"
  >
    This is a text
  </a>
</span>
`;

exports[`When using snapshots on an anchor typed Button Should render with an element children 1`] = `
<span
  className="sc-bdVaJa lbJuWU"
>
  <a
    className="btn shape-round btn-default btn-md sc-htpNat jTMgHP"
    disabled={false}
    role="button"
  >
    <div />
  </a>
</span>
`;
